### PR TITLE
Adds support for limited version of managed blocking that aborts if ther...

### PR DESCRIPTION
**Pull request by viktorklang**
...e aren't enough **not blocked** threads available in the pool

Do not merge yet! :)

### CODE REVIEW

File: akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala [hunk 1]
```diff
@@ -46,13 +47,45 @@ object ThreadPoolConfig {
...
+      delegate.newThread(new Runnable with BlockContext {
+        override def run(): Unit = BlockContext.withBlockContext(this) {
+          incrementAndGet()
+          try runnable.run() finally { decrementAndGet() }
+        }
+
+        override def blockOn[T](thunk: ⇒ T)(implicit permission: CanAwait): T = {
+          @scala.annotation.tailrec def canBlock(): Boolean = {
+            val pre = get()
+            if (pre > 0 && pre > minUnblockedThreads) compareAndSet(pre, pre - 1) || canBlock()
```
Comment (rkuhn) [inline akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala]: `minUnblockedThreads > 0`, which means that `pre > 0` is superfluous

Comment (rkuhn) [inline akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala]: Also: no loop necessary if you just `decrementAndGet`, correcting if the result was negative.

File: akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala [hunk 2]
```diff
@@ -46,13 +47,45 @@ object ThreadPoolConfig {
...
+            else {
+              val tpe = threadPoolExecutor
+              (tpe ne null) && tpe.prestartCoreThread() && canBlock() // Retry if we have a TPE and we booted a new Core Thread
```
Comment (rkuhn) [inline akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala]: So this is the place where we’d also need to allow “going negative” if there is headroom towards max-pool-threads?

**Comment by rkuhn** [comment]
I’d like to either close or merge this within a week, which means that in the absence of brilliant ideas we’ll auto-close.

File: akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala (hunk 1 above)
Comment (viktorklang) [inline akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala]: What makes you say `minUnblockedThreads > 0`?

Comment (viktorklang) [inline akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala]: @rkuhn I find compensating for negatives leads to harder to maintain code than TTAS but YMMV

**Comment by viktorklang** [comment]
I'm OK with either, I just thought the idea of managed blocking for TPE was a novelty.

File: akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala (hunk 1 above)
Comment (rkuhn) [inline akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala]: The condition upon which the involvement of this class depends is `if (minUnblockedThreads > 0)`, see [here](https://github.com/akka/akka/pull/16421/files#diff-410abf8aac6a89a01b05db2cbcf1765cR113).

Comment (viktorklang) [inline akka-actor/src/main/scala/akka/dispatch/ThreadPoolBuilder.scala]: That's a use-site, not an invariant within the implementation.

**Comment by rkuhn** [comment]
Yes, it is, but unless it works with max > core I’d say it is a rather … queer … novelty ;-)

**Comment by viktorklang** [comment]
@rkuhn Something is better than nothing ;)

**Comment by viktorklang** [comment]
@rkuhn Sorry for being indirect: You have my full support in the decision to either close or merge this rather PoC "feature". I'd lean towards not merging.

**Pull request closed without merging**